Resolve entities defined in DOCTYPE recursively.

jgm opened the pull request:
This puts a limit on recursion depth (10) to avoid loops and other attacks.

Fixes #103.

k0ral commented on xml-conduit/src/Text/XML/Stream/Parse.hs:
```diff
@@ -227,13 +224,31 @@ tokenToEvent _ es n (TokenEndElement name) =
+                -> [Content]
+                -> [Content]
+resolveEntities depth ps es (ContentEntity e : cs)
+    | depth < 10  -- avoid malicious recursive entity expansion
```
What about adding a field in `ParseSettings` to control the maximum depth, with default value set to 10 ?

jgm replied:
I could do that if you want, but it does mean an API change -- do you really want that?

jgm replied:
Or we could just set this to a higher number, such that no realistic case would exceed it.
(I think 10 already might be high enough, but what about 100?)

jgm replied:
Another option is to avoid having a limit on recursion depth, but to keep track of the entities we've already expanded in a Set.  If we encounter `bar` in the process of recursively expanding `bar`, then we stop.  That would prevent loops without hard-coding any number.  I'd favor this approach I think.

Another issue that was raised in #103 was the billion laughs attack, e.g.:
```xml
<?xml version="1.0"?>
<!DOCTYPE lolz [
 <!ENTITY lol "lol">
 <!ELEMENT lolz (#PCDATA)>
 <!ENTITY lol1 "&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;">
 <!ENTITY lol2 "&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;">
 <!ENTITY lol3 "&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;">
 <!ENTITY lol4 "&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;">
 <!ENTITY lol5 "&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;">
 <!ENTITY lol6 "&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;">
 <!ENTITY lol7 "&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;">
 <!ENTITY lol8 "&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;">
 <!ENTITY lol9 "&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;">
]>
<lolz>&lol9;</lolz>
```
But note that even setting recursion depth to 10 won't prevent this.  I'm not sure preventing this should be a goal of an XML parsing library, since this is just valid XML with a very long expansion.  But if you were worried about this, we could trigger an exception when the expansion of an entity exceeds some fixed length?

jgm replied:
[This article](https://docs.microsoft.com/en-us/archive/msdn-magazine/2009/november/xml-denial-of-service-attacks-and-defenses) recommends having an option to disable DTD parsing and another option to set the maximum number of characters for entity expansion.  That sounds like a reasonable approach; it would require two additional fields on parser options.

k0ral replied:
> I could do that if you want, but it does mean an API change -- do you really want that?

Constructor of `ParseSettings` isn't exported (see [Text.XML.Stream.Parse#77](https://github.com/snoyberg/xml/blob/master/xml-conduit/src/Text/XML/Stream/Parse.hs#L77)), so it should be backward compatible, unless I am mistaken ?

Even if it weren't, the alternative of having a hardcoded value sounds worse to me, from a user standpoint.

> Another option is to avoid having a limit on recursion depth, but to keep track of the entities we've already expanded in a Set. If we encounter bar in the process of recursively expanding bar, then we stop.

Sounds correct to me, but considering the extra cost in time and memory, in this context, a simpler recursion-depth-based heuristic sounds like a nice tradeoff.

> I'm not sure preventing this should be a goal of an XML parsing library, since this is just valid XML with a very long expansion.

I am not sure it should either, but simple safeguards that don't compromise performance, usability and maintainability are probably fair game. Having a maximum number of characters after expansion seems to qualify.

jgm commented:
OK, I've changed this, adding a new adjustable parameter for the max size of any single entity expansion, and better code for avoiding loops. I've also added more tests.

k0ral commented on xml-conduit/src/Text/XML/Stream/Parse.hs:
```diff
@@ -227,13 +227,48 @@ tokenToEvent _ es n (TokenEndElement name) =
+          -- to avoid infinite loops in recursive expansion:
+          case foldr (goent (filter (\(x,_) -> x /= e) es)) (Just ([], 0)) xs of
+            Just (cs, _) -> Just cs
+            Nothing -> Nothing
```
This:
```haskell
case something of
  Just (cs, _) -> Just cs
  Nothing -> Nothing
```
... can be written in an arguably more readable way:
```haskell
fst <$> something
```

k0ral commented on xml-conduit/src/Text/XML/Stream/Parse.hs:
```diff
@@ -227,13 +227,48 @@ tokenToEvent _ es n (TokenEndElement name) =
+  goent es c@(ContentEntity e) (Just (cs, size))
+    = case expandEntity es e of
+        Just xs -> foldr (goent es) (Just (cs, size)) xs
+        Nothing -> Nothing
```
Similarly, I believe this can be simplified into:
```haskell
foldr (goent es) (Just (cs, size)) <$> expandEntity es e
```

jgm replied:
Actually it needs to be
```
expandEntity es e >>= foldr etc.
```
I'll write it that way if you prefer that style.

k0ral replied:
Right, my bad.

k0ral commented on xml-conduit/test/unit.hs:
```diff
@@ -736,6 +740,37 @@ resolvedInline = do
+resolvedInlineLoop = do
+    res <- return $ Res.parseLBS Res.def
+           "<!DOCTYPE foo [<!ENTITY bim \"&bim;\">]><foo>&bim;</foo>"
+    Left (toException $ Res.UnresolvedEntityException (Set.fromList ["bim"]))
```
[Optional] We may want to provide more info in the exception structure, to distinguish between:
- undefined entities
- recursive entities
- entities for which max expansion size has been reached

jgm replied:
Currently we don't raise any exception in `resolveEntities` itself (indeed, it's just a pure function and we can't raise the exception directly there). Instead, when an entity cannot be successfully resolved, either because it's not defined, or because it loops, or because its expansion is too large, we just pass through a `ContentEntity` token.  The exception gets raised at a later stage when this token is encountered (by some code I haven't touched, I guess `fromEvents` and `fromXMLElement` in Text.XML.).

To allow that code to produce different exceptions for these different cases, we'd need to have an additional parameter on `ContentEntity` (if that's just an internal type then maybe that's okay).

Maybe something like

```
ContentEntity EntityStatus Text
```
where
```
data EntityStatus = Unresolved | Loop | ExceedsMaxSize
```

jgm replied:
If you want to go this route, you'd need to define for me what exceptions you want to produce from `fromEvents` for each of these cases.

k0ral replied:
Indeed, it seems it would require many more changes, that are beyond the scope of this pull-request. I'll have a look after merging it.

k0ral reviewed (commented):
Sorry for the delay in reviewing.

jgm commented on xml-conduit/src/Text/XML/Stream/Parse.hs:
```diff
@@ -227,13 +227,48 @@ tokenToEvent _ es n (TokenEndElement name) =
+        Right xs    ->
+          -- we delete e from the entity map in resolving its contents,
+          -- to avoid infinite loops in recursive expansion:
+          case foldr (goent (filter (\(x,_) -> x /= e) es)) (Just ([], 0)) xs of
```
I thought about doing that.  The main impediment is that it would be awkward to do this without changing the `TokenDoctype` constructor
```
| TokenDoctype Text (Maybe ExternalID) [(Text, Text)]
```
which is exported, so this would be a breaking API change. If that's okay with you, I could go ahead and make the change. Another alternative is to keep the association list in `TokenDoctype`, but convert it to a map either in `resolveEntities` or before calling `tokenToEvent` (which isn't exported and could have its signature change).

I was reluctant to make any of these changes, because I'm not really sure what the performance impact would be. I imagine most doctypes have no more than 20 internal entities, and it's quite possible that the overhead of converting the list to a map would swamp the savings from the lookup/delete.  I'm just not sure.  Is there a benchmark suite for this library?

k0ral replied:
Fair points. There is currently no benchmark suite for this library, I might look into that after merging this pull-request.

k0ral commented:
Squashed and merged, thank you !

jgm commented:
Great, thank you for the quick merge.  When do you expect a new release of xml-conduit?  I am looking forward to switching skylighting from HXT, but this awaits these fixes.

k0ral commented:
Released as [1.9.1.0](https://hackage.haskell.org/package/xml-conduit-1.9.1.0).
